@typescript/vfs type declaration file is broken

I'd like to reopen #2742, it isn't website related but it concerns the `@typescript/vfs`. 

> `@typescript/vfs`'s `index.d.ts` file makes references to things that don't exist in it's scope.
> 
> **Expected behavior:**
> 
> `> tsc --noEmit` doesn't give any error for this module.
> 
> **Actual behavior:**
> 
> There are 3 errors:
> 
> ```shell
> > tsc --noEmit
> 
> node_modules/.pnpm/@typescript+vfs@1.4.0/node_modules/@typescript/vfs/dist/index.d.ts:57:140 - error TS2307: Cannot find module 'lz-string' or its corresponding type declarations.
> 
> 57 export declare const createDefaultMapFromCDN: (options: CompilerOptions, version: string, cache: boolean, ts: TS, lzstring?: typeof import("lz-string"), fetcher?: typeof fetch, storer?: typeof localStorage) => Promise<Map<string, string>>;
>                                                                                                                                               ~~~~~~~~~~~
> 
> node_modules/.pnpm/@typescript+vfs@1.4.0/node_modules/@typescript/vfs/dist/index.d.ts:57:171 - error TS2552: Cannot find name 'fetch'. Did you mean 'fetcher'?
> 
> 57 export declare const createDefaultMapFromCDN: (options: CompilerOptions, version: string, cache: boolean, ts: TS, lzstring?: typeof import("lz-string"), fetcher?: typeof fetch, storer?: typeof localStorage) => Promise<Map<string, string>>;
>                                                                                                                                                                              ~~~~~
> 
>   node_modules/.pnpm/@typescript+vfs@1.4.0/node_modules/@typescript/vfs/dist/index.d.ts:57:154
>     57 export declare const createDefaultMapFromCDN: (options: CompilerOptions, version: string, cache: boolean, ts: TS, lzstring?: typeof import("lz-string"), fetcher?: typeof fetch, storer?: typeof localStorage) => Promise<Map<string, string>>;
>                                                                                                                                                                 ~~~~~~~~~~~~~~~~~~~~~~
>     'fetcher' is declared here.
> 
> node_modules/.pnpm/@typescript+vfs@1.4.0/node_modules/@typescript/vfs/dist/index.d.ts:57:194 - error TS2304: Cannot find name 'localStorage'.
> 
> 57 export declare const createDefaultMapFromCDN: (options: CompilerOptions, version: string, cache: boolean, ts: TS, lzstring?: typeof import("lz-string"), fetcher?: typeof fetch, storer?: typeof localStorage) => Promise<Map<string, string>>;
>                                                                                                                                                                                                     ~~~~~~~~~~~~
> 
> 
> Found 3 errors in the same file, starting at: node_modules/.pnpm/@typescript+vfs@1.4.0/node_modules/@typescript/vfs/dist/index.d.ts:57
> 
>  ELIFECYCLE  Command failed with exit code 2.
> ```


`typeof fetch` is no-longer an issue since Node18, but others still are.

--- a/.changeset/silver-kangaroos-exist.md
+++ b/.changeset/silver-kangaroos-exist.md
@@ -1,0 +1,5 @@
+---
+"@typescript/vfs": patch
+---
+
+Don't depend on DOM types in createDefaultMapFromCDN

--- a/packages/typescript-vfs/package.json
+++ b/packages/typescript-vfs/package.json
@@ -51,7 +51,6 @@
     "jest": "^29.5.0",
     "jest-environment-jsdom": "^29.5.0",
     "jest-watch-typeahead": "^2.2.2",
-    "lz-string": "^1.5.0",
     "ts-jest": "^29.0.5",
     "tslib": "^2.6.2",
     "typescript": "*"

--- a/packages/typescript-vfs/src/index.ts
+++ b/packages/typescript-vfs/src/index.ts
@@ -6,13 +6,24 @@ type CompilerHost = import("typescript").CompilerHost
 type SourceFile = import("typescript").SourceFile
 type TS = typeof import("typescript")
 
+type FetchLike = (url: string) => Promise<{ json(): Promise<any>; text(): Promise<string> }>
+
+interface LocalStorageLike {
+  getItem(key: string): string | null
+  setItem(key: string, value: string): void
+  removeItem(key: string): void
+}
+
+declare var localStorage: LocalStorageLike | undefined;
+declare var fetch: FetchLike | undefined;
+
 let hasLocalStorage = false
 try {
   hasLocalStorage = typeof localStorage !== `undefined`
 } catch (error) { }
 
 const hasProcess = typeof process !== `undefined`
-const shouldDebug = (hasLocalStorage && localStorage.getItem("DEBUG")) || (hasProcess && process.env.DEBUG)
+const shouldDebug = (hasLocalStorage && localStorage!.getItem("DEBUG")) || (hasProcess && process.env.DEBUG)
 const debugLog = shouldDebug ? console.log : (_message?: any, ..._optionalParams: any[]) => ""
 
 export interface VirtualTypeScriptEnvironment {
@@ -297,14 +308,6 @@ export interface LZString {
   decompressFromUTF16(compressed: string): string
 }
 
-export type FetchLike = (url: string) => Promise<{ json(): Promise<any>; text(): Promise<string> }>
-
-export interface LocalStorageLike {
-  getItem(key: string): string | null
-  setItem(key: string, value: string): void
-  removeItem(key: string): void
-}
-
 /**
  * Create a virtual FS Map with the lib files from a particular TypeScript
  * version based on the target, Always includes dom ATM.
@@ -326,7 +329,7 @@ export const createDefaultMapFromCDN = (
   fetcher?: FetchLike,
   storer?: LocalStorageLike
 ) => {
-  const fetchlike = fetcher || fetch
+  const fetchlike = fetcher || fetch!
   const fsMap = new Map<string, string>()
   const files = knownLibFilesForCompilerOptions(options, ts)
   const prefix = `https://playgroundcdn.typescriptlang.org/cdn/${version}/typescript/lib/`
@@ -353,7 +356,7 @@ export const createDefaultMapFromCDN = (
 
   // A localstorage and lzzip aware version of the lib files
   function cached() {
-    const storelike = storer || localStorage
+    const storelike = storer || localStorage!
 
     const keys = Object.keys(storelike)
     keys.forEach(key => {

--- a/packages/typescript-vfs/src/index.ts
+++ b/packages/typescript-vfs/src/index.ts
@@ -292,6 +292,19 @@ export const addAllFilesFromFolder = (map: Map<string, string>, workingDir: stri
 export const addFilesForTypesIntoFolder = (map: Map<string, string>) =>
   addAllFilesFromFolder(map, "node_modules/@types")
 
+export interface LZString {
+  compressToUTF16(input: string): string
+  decompressFromUTF16(compressed: string): string
+}
+
+export type FetchLike = (url: string) => Promise<{ json(): Promise<any>; text(): Promise<string> }>
+
+export interface LocalStorageLike {
+  getItem(key: string): string | null
+  setItem(key: string, value: string): void
+  removeItem(key: string): void
+}
+
 /**
  * Create a virtual FS Map with the lib files from a particular TypeScript
  * version based on the target, Always includes dom ATM.
@@ -309,9 +322,9 @@ export const createDefaultMapFromCDN = (
   version: string,
   cache: boolean,
   ts: TS,
-  lzstring?: typeof import("lz-string"),
-  fetcher?: typeof fetch,
-  storer?: typeof localStorage
+  lzstring?: LZString,
+  fetcher?: FetchLike,
+  storer?: LocalStorageLike
 ) => {
   const fetchlike = fetcher || fetch
   const fsMap = new Map<string, string>()

--- a/packages/typescript-vfs/tsconfig.json
+++ b/packages/typescript-vfs/tsconfig.json
@@ -5,7 +5,7 @@
   "compilerOptions": {
     "target": "ES2015",
     "module": "esnext",
-    "lib": ["dom", "esnext"],
+    "lib": ["esnext"],
     "importHelpers": true,
     "declaration": true,
     "sourceMap": true,

